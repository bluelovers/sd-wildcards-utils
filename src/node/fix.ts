import { IWildcardsYAMLMapRoot } from '../types';

export function _fixYAMLMapCommentBefore(node: IWildcardsYAMLMapRoot)
{
	if (node.commentBefore?.length)
	{
		const pairKey = node.items[0]?.key;

		if (pairKey)
		{
			let msg = node.commentBefore;
			let key: 'comment' | 'commentBefore' = 'commentBefore';
			if (pairKey[key] && msg !== pairKey[key])
			{
				msg = `${pairKey[key]}\n${msg}`;
			}

			pairKey[key] = msg;
			node.commentBefore = void 0;
		}
	}
}
