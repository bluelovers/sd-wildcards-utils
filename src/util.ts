import { array_unique_overwrite } from 'array-hyper-unique';
import {
	IMatchDynamicPromptsWildcardsEntry,
	IVisitPathsNode,
	IYamlNodeTypeSymbol
} from './types';

export const SYMBOL_YAML_NODE_TYPE_ALIAS = Symbol.for('yaml.alias');
export const SYMBOL_YAML_NODE_TYPE_DOC = Symbol.for('yaml.document');
export const SYMBOL_YAML_NODE_TYPE_MAP = Symbol.for('yaml.map');
export const SYMBOL_YAML_NODE_TYPE_PAIR = Symbol.for('yaml.pair');
export const SYMBOL_YAML_NODE_TYPE_SCALAR = Symbol.for('yaml.scalar');
export const SYMBOL_YAML_NODE_TYPE_SEQ = Symbol.for('yaml.seq');
export const SYMBOL_YAML_NODE_TYPE = Symbol.for('yaml.node.type');

export const RE_DYNAMIC_PROMPTS_WILDCARDS = /(?<!#[^\n]*)__([&~!@])?([*\w\/_\-]+)(\([^\n#]+\))?__/

/**
 * for `matchAll`
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll
 */
export const RE_DYNAMIC_PROMPTS_WILDCARDS_GLOBAL = new RegExp(RE_DYNAMIC_PROMPTS_WILDCARDS, RE_DYNAMIC_PROMPTS_WILDCARDS.flags + 'g')

export const RE_WILDCARDS_NAME = /^[\w\-_\/]+$/

/**
 * Checks if the input string matches the dynamic prompts wildcards pattern.
 *
 * @param input - The input string to check.
 * @returns A boolean indicating whether the input string matches the pattern.
 *
 * @remarks
 * This function uses the `matchDynamicPromptsWildcards` function to perform the check.
 * It returns `true` if the input string is a full match, and `false` otherwise.
 *
 * @example
 * ```typescript
 * const input1 = "__season_clothes(season=winter)__";
 * console.log(isDynamicPromptsWildcards(input1)); // Output: true
 *
 * const input2 = "__season_clothes(season=__season_clothes__)__";
 * console.log(isDynamicPromptsWildcards(input2)); // Output: true
 *
 * const input3 = "This is not a wildcards pattern";
 * console.log(isDynamicPromptsWildcards(input3)); // Output: false
 * ```
 */
export function isDynamicPromptsWildcards(input: string): boolean
{
	return matchDynamicPromptsWildcards(input).isFullMatch;
}

/**
 * Matches the input string against the dynamic prompts wildcards pattern.
 *
 * @see https://github.com/adieyal/sd-dynamic-prompts/blob/main/docs/SYNTAX.md
 *
 * @param input - The input string to match.
 * @returns An object containing the matched groups or `null` if no match is found.
 *
 * @remarks
 * This function uses the `RE_DYNAMIC_PROMPTS_WILDCARDS` regular expression to perform the match.
 * The returned object contains the following properties:
 * - `name`: The name extracted from the input string.
 * - `variables`: The variables extracted from the input string.
 * - `keyword`: The keyword extracted from the input string.
 * - `source`: The original matched source string.
 * - `isFullMatch`: A boolean indicating whether the input string is a full match.
 *
 * @example
 * ```typescript
 * const input = "\_\_season_clothes(season=winter)\_\_";
 * const result = matchDynamicPromptsWildcards(input);
 * console.log(result);
 * // Output: { name: 'season_clothes', variables: '(season=winter)', keyword: undefined, source: '\__season_clothes(season=winter)\__', isFullMatch: true }
 * ```
 *
 * @example
 * __season_clothes(season=winter)__
 * __season_clothes(season=__season_clothes__)__
 * __season_clothes(season=!__season_clothes__)__
 *
 * __season_clothes(season=__@season_clothes__)__
 * __season_clothes(season=__~season_clothes__)__
 *
 * __@season_clothes(season=__season_clothes__)__
 * __~season_clothes(season=__season_clothes__)__
 *
 * __season_clothes(season={summer|autumn|winter|spring})__
 * __season_clothes(season=!{summer|autumn|winter|spring})__
 *
 * __season_clothes(season={@summer|autumn|winter|spring})__
 * __season_clothes(season={!summer|autumn|winter|spring})__
 *
 * __season_clothes(season=)__
 */
export function matchDynamicPromptsWildcards(input: string)
{
	const m = input.match(RE_DYNAMIC_PROMPTS_WILDCARDS);
	return _matchDynamicPromptsWildcardsCore(m, input);
}

export function _matchDynamicPromptsWildcardsCore(m: RegExpMatchArray,
	input?: string,
): IMatchDynamicPromptsWildcardsEntry
{
	if (!m) return null;

	let [source, keyword, name, variables] = m;

	return {
		name,
		variables,
		keyword,
		source,
		isFullMatch: source === (input ?? m.input),
		isStarWildcards: name.includes('*'),
	}
}

/**
 * Generator function that matches all occurrences of the dynamic prompts wildcards pattern in the input string.
 */
export function* matchDynamicPromptsWildcardsAllGenerator(input: string)
{
	const ls = input.matchAll(RE_DYNAMIC_PROMPTS_WILDCARDS_GLOBAL);

	for (let m of ls)
	{
		yield _matchDynamicPromptsWildcardsCore(m, input);
	}
}

/**
 * Converts the generator function `matchDynamicPromptsWildcardsAllGenerator` into an array.
 */
export function matchDynamicPromptsWildcardsAll(input: string, unique?: boolean)
{
	const arr = [...matchDynamicPromptsWildcardsAllGenerator(input)] as IMatchDynamicPromptsWildcardsEntry[];

	return unique ? array_unique_overwrite(arr) : arr
}

/**
 * Checks if the given name is a valid Wildcards name.
 *
 * @param name - The name to check.
 * @returns A boolean indicating whether the name is valid.
 *
 * @remarks
 * A valid Wildcards name should:
 * - Only contain alphanumeric characters, hyphens, or underscores.
 * - Not start or end with an underscore.
 * - Not contain consecutive underscores.
 *
 * @example
 * ```typescript
 * const name1 = "season_clothes";
 * console.log(isWildcardsName(name1)); // Output: true
 *
 * const name2 = "_season_clothes";
 * console.log(isWildcardsName(name2)); // Output: false
 *
 * const name3 = "season_clothes_";
 * console.log(isWildcardsName(name3)); // Output: false
 *
 * const name4 = "season__clothes";
 * console.log(isWildcardsName(name4)); // Output: false
 *
 * const name5 = "season-clothes";
 * console.log(isWildcardsName(name5)); // Output: true
 * ```
 */
export function isWildcardsName(name: string): boolean
{
	return RE_WILDCARDS_NAME.test(name) && !/__|[_\/]$|^[_\/]|\/\//.test(name)
}

export function assertWildcardsName(name: string)
{
	if (isWildcardsName(name))
	{
		throw new SyntaxError(`Invalid Wildcards Name Syntax: ${name}`)
	}
}

export function convertWildcardsNameToPaths(name: string)
{
	return name.split('/');
}

export function isWildcardsPathSyntx(path: string): path is `__${string}__`
{
	return RE_DYNAMIC_PROMPTS_WILDCARDS.test(path)
}

export function wildcardsPathToPaths(path: string)
{
	if (isWildcardsPathSyntx(path))
	{
		path = matchDynamicPromptsWildcards(path).name
	}

	return convertWildcardsNameToPaths(path);
}

export function getNodeTypeSymbol(node: IVisitPathsNode): IYamlNodeTypeSymbol
{
	// @ts-ignore
	return node?.[SYMBOL_YAML_NODE_TYPE]
}

export function _getNodeTypeCore(sym: IYamlNodeTypeSymbol)
{
	try
	{
		return Symbol.keyFor(sym)
	}
	catch (e)
	{

	}
}

export function getNodeType(node: IVisitPathsNode)
{
	return _getNodeTypeCore(getNodeTypeSymbol(node))
}

export function isSameNodeType(a: IVisitPathsNode, b: IVisitPathsNode)
{
	const s = getNodeTypeSymbol(a);

	return s && getNodeTypeSymbol(b) === s;
}
