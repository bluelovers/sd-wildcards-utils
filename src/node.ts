/**
 * Created by user on 2025/9/14.
 */

import { isDocument, isMap, isSeq, Scalar } from 'yaml';
import {
	IWildcardsYAMLDocument,
	IWildcardsYAMLPair,
	IWildcardsYAMLScalar,
	IYAMLCollectionNode,
	IYAMLNodeBaseLike,
} from './types';
import { isUnset } from './util';
import { isWildcardsYAMLScalar } from './node-is';

export interface INodeCopyMergeOptions
{
	overwrite?: boolean,
	merge?: boolean | number,
}

/**
 * Preserve comments from the original key (if any)
 */
export function _nodeCopyMergeCommentCore(node: IYAMLNodeBaseLike, nodeOld: IYAMLNodeBaseLike, key: 'commentBefore' | 'comment', opts: INodeCopyMergeOptions)
{
	const oldValue = nodeOld[key];
	const curValue = node[key];

	if (oldValue !== curValue)
	{
		if (opts.merge && curValue?.length)
		{
			node[key] = (opts.merge as number) > 1 ? `${oldValue}\n\n${curValue}` : `${curValue}\n\n${oldValue}`
		}
		else if (opts.overwrite || opts.merge && oldValue?.length)
		{
			node[key] = oldValue
		}
		else
		{
			node[key] ??= oldValue
		}
	}
}

export function nodeHasComment(node: IYAMLNodeBaseLike)
{
	return node && (node.commentBefore?.length || node.comment?.length);
}

export function _copyMergeNodeCore<T extends IYAMLNodeBaseLike, R extends IYAMLNodeBaseLike>(node: T, nodeOld: R, opts: INodeCopyMergeOptions)
{
	_nodeCopyMergeCommentCore(node, nodeOld, 'commentBefore', opts);
	_nodeCopyMergeCommentCore(node, nodeOld, 'comment', opts);
}

export function copyMergeScalar<T extends IWildcardsYAMLScalar | Scalar>(node: T, nodeOld: unknown, opts?: INodeCopyMergeOptions)
{
	if (!isWildcardsYAMLScalar(node) || !isWildcardsYAMLScalar(nodeOld))
	{
		throw new TypeError('node and nodeOld must be Scalar')
	}

	opts ??= {};

	_copyMergeNodeCore(node, nodeOld, opts);

	if (!isUnset(nodeOld.spaceBefore) && (isUnset(node.spaceBefore) || (opts.overwrite || opts.merge)))
	{
		node.spaceBefore = nodeOld.spaceBefore
	}

	node.value ??= nodeOld.value
}

export function nodeGetItems<T extends IWildcardsYAMLPair>(node: IYAMLCollectionNode): T[]
{
	if (isDocument(node))
	{
		return (node as IWildcardsYAMLDocument).contents?.items as any[]
	}
	else if (isSeq(node) || isMap(node))
	{
		return node.items as any[]
	}
}
