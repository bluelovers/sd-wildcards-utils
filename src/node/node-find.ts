import { IVisitPathsNodeList, IWildcardsYAMLPair, IYAMLCollectionNode } from '../types';
import { isDocument, isPair, isSeq, YAMLMap } from 'yaml';
import { findPair } from 'yaml/util';
import { nodeGetItems } from './node';

export function findUpParentNodes(nodeList: IVisitPathsNodeList)
{
	let _cache: IWildcardsYAMLPair[] = [];

	for (let i = nodeList.length - 1; i >= 0; i--)
	{
		const node = nodeList[i];

		if (isSeq(node))
		{
			continue;
		}

		if (isPair(node))
		{
			_cache.unshift((node as IWildcardsYAMLPair))
		}
		else if (isDocument(node))
		{
			//_cache.unshift((node as any))
		}
	}

	return _cache;
}

export function findUpParentNodesNames(nodeList: IVisitPathsNodeList)
{
	let _cache: string[] = [];

	for (let i = nodeList.length - 1; i >= 0; i--)
	{
		const node = nodeList[i];

		if (isSeq(node))
		{
			continue;
		}

		if (isPair(node))
		{
			_cache.unshift((node as IWildcardsYAMLPair).key.value)
		}
	}

	return _cache;
}

export function _nodeGetInPairCore(node: IYAMLCollectionNode, key: unknown)
{
	const items = nodeGetItems(node);

	return items && findPair(items, key) as IWildcardsYAMLPair
}

export function nodeGetInPair(node: IYAMLCollectionNode, paths: readonly unknown[])
{
	if (paths.length === 1)
	{
		return _nodeGetInPairCore(node, paths[0])
	}
	else if (paths.length > 0)
	{
		const parent = node.getIn(paths.slice(0, -1)) as YAMLMap;

		return _nodeGetInPairCore(parent, paths[paths.length - 1])
	}
}

export function nodeGetInPairAll(node: IYAMLCollectionNode, paths: readonly unknown[])
{
	let list: IWildcardsYAMLPair[] = [];

	let cur: IYAMLCollectionNode = node;

	for (const key of paths)
	{
		let pair = nodeGetInPair(cur, [key]);

		if (pair)
		{
			list.push(pair);

			cur = pair.value
		}
		else
		{
			break;
		}
	}

	return list
}
