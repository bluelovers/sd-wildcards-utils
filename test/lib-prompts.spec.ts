//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { join, normalize } from 'upath2';
import { globSync } from 'node:fs';
import { __ROOT_TEST_FIXTURES, __ROOT_TEST_SNAPSHOTS_FILE } from './__root';
import { readFileSync } from 'fs-extra';
import {
	parseWildcardsYaml,
	_checkValue,
	IWildcardsYAMLDocumentParsed,
	normalizeDocument,
	stringifyWildcardsYamlData,
} from '../src/index';
import { globSync2 } from './script/lib/util';

beforeAll(async () =>
{

});

describe(`_checkValue`, () => {

	describe(`valid`, () => {

		test.each([
			'2b_\(nier:automata\)_\(cosplay\)',
			'purple_gray',

			`, penis under another's clothes`,

			`{ penis under another's clothes|}`,

			`{, penis under another's clothes}`,

			`{, penis under another's clothes|}`,
			`(naizuri{, penis under another's clothes|}:1.3)`,
		])(`%j`, (input) => {

			let actual = _checkValue(input);

			expect(actual).toBeUndefined();
		})

	})

	describe(`invalid`, () => {

		test.each([
			' __lazy-wildcards/subject/env-elem/stairs/prompts_ ',
			' __lazy-wildcards/subject/__env-elem/stairs/prompts__ ',
			' {__lazy-wildcards/subject/env-elem/stairs/prompts_} ',
			' _lazy-wildcards/subject/env-elem/stairs/prompts__ ',
			' __lazy-wildcards/subject/env-elem__/stairs/prompts__ ',
			' {_lazy-wildcards/subject/env-elem/stairs/prompts__} ',
			`(cum
{, __1/subject/costume-elem/cum/costume-elem2__|}
{, __2/subject/costume-elem/cum/costume-elem__|
{, __3/subject/costume-elem/cum-base/prompts__{0.3:::{1.2|1.3|1.4}|}|})
`,

			`__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v=!xxx)__`,
			`__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v={x})__`,
			`__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v=\${x})__`,

		])(`%j`, (input) => {

			let actual = _checkValue(input);

			expect(actual).not.toBeUndefined();
			expect(actual).toMatchSnapshot();
		})

	})

});

describe(`prompts:error`, () =>
{
	test.each(globSync2([
		`prompts-bad/**.yaml`,
	], {
		cwd: __ROOT_TEST_FIXTURES,
	}))('%j', (file) =>
	{

		const outPath = join(
			__ROOT_TEST_SNAPSHOTS_FILE,
			'stringifyWildcardsYamlData',
		);

		let source = readFileSync(join(__ROOT_TEST_FIXTURES, file));

		let yaml: IWildcardsYAMLDocumentParsed;

		expect(() =>
		{
			yaml = parseWildcardsYaml(source, {
				allowMultiRoot: true,
			});

			stringifyWildcardsYamlData(yaml);
		}).toThrowErrorMatchingSnapshot()
	});

})
