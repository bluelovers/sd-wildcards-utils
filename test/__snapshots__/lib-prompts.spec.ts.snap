// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`_checkValue invalid " __lazy-wildcards/subject/__env-elem/stairs/prompts__ " 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "/__" in value near "ldcards/subject/__env-elem/stairs"",
  "index": 25,
  "match": "/__",
  "near": "ldcards/subject/__env-elem/stairs",
  "value": " __lazy-wildcards/subject/__env-elem/stairs/prompts__ ",
}
`;

exports[`_checkValue invalid " __lazy-wildcards/subject/env-elem/stairs/prompts_ " 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "_ " in value near "/stairs/prompts_ "",
  "index": 49,
  "match": "_ ",
  "near": "/stairs/prompts_ ",
  "value": " __lazy-wildcards/subject/env-elem/stairs/prompts_ ",
}
`;

exports[`_checkValue invalid " __lazy-wildcards/subject/env-elem__/stairs/prompts__ " 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "__/" in value near "ubject/env-elem__/stairs/prompts_"",
  "index": 34,
  "match": "__/",
  "near": "ubject/env-elem__/stairs/prompts_",
  "value": " __lazy-wildcards/subject/env-elem__/stairs/prompts__ ",
}
`;

exports[`_checkValue invalid " _lazy-wildcards/subject/env-elem/stairs/prompts__ " 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] " _" in value near " _lazy-wildcards/"",
  "index": 0,
  "match": " _",
  "near": " _lazy-wildcards/",
  "value": " _lazy-wildcards/subject/env-elem/stairs/prompts__ ",
}
`;

exports[`_checkValue invalid " {__lazy-wildcards/subject/env-elem/stairs/prompts_} " 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "_}" in value near "/stairs/prompts_} "",
  "index": 50,
  "match": "_}",
  "near": "/stairs/prompts_} ",
  "value": " {__lazy-wildcards/subject/env-elem/stairs/prompts_} ",
}
`;

exports[`_checkValue invalid " {_lazy-wildcards/subject/env-elem/stairs/prompts__} " 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "{_" in value near " {_lazy-wildcards/"",
  "index": 1,
  "match": "{_",
  "near": " {_lazy-wildcards/",
  "value": " {_lazy-wildcards/subject/env-elem/stairs/prompts__} ",
}
`;

exports[`_checkValue invalid "(cum\\n{, __1/subject/costume-elem/cum/costume-elem2__|}\\n{, __2/subject/costume-elem/cum/costume-elem__|\\n{, __3/subject/costume-elem/cum-base/prompts__{0.3:::{1.2|1.3|1.4}|}|})\\n" 1`] = `
{
  "error": "Invalid Error [UNKNOWN]: TypeError: this.result.simple.push is not a function",
  "value": "(cum
{, __1/subject/costume-elem/cum/costume-elem2__|}
{, __2/subject/costume-elem/cum/costume-elem__|
{, __3/subject/costume-elem/cum-base/prompts__{0.3:::{1.2|1.3|1.4}|}|})
",
}
`;

exports[`_checkValue invalid "__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v=!xxx)__" 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "(v=!" in value near "/fn/rocket_tits(v=!xxx)__"",
  "index": 54,
  "match": "(v=!",
  "near": "/fn/rocket_tits(v=!xxx)__",
  "value": "__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v=!xxx)__",
}
`;

exports[`_checkValue invalid "__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v=\${x})__" 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "(v=$" in value near "/fn/rocket_tits(v=\${x})__"",
  "index": 54,
  "match": "(v=$",
  "near": "/fn/rocket_tits(v=\${x})__",
  "value": "__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v=\${x})__",
}
`;

exports[`_checkValue invalid "__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v={x})__" 1`] = `
{
  "error": "Invalid Syntax [UNSAFE_SYNTAX] "(v={" in value near "/fn/rocket_tits(v={x})__"",
  "index": 54,
  "match": "(v={",
  "near": "/fn/rocket_tits(v={x})__",
  "value": "__costume-ethnicity-breasts/tits-rocket/fn/rocket_tits(v={x})__",
}
`;

exports[`prompts:error "prompts-bad/not-matched_001.yaml" 1`] = `"Invalid Syntax [BRACKET] Unable to parse. Unclosed Bracket detected near "it_commoner_any__". key: 0, node: cdfy/outfit_commoner_any__"`;

exports[`prompts:error "prompts-bad/not-matched_002.yaml" 1`] = `"Invalid Syntax [BRACKET] Unable to parse. Unclosed Bracket detected near "__cdfy/outfit_commoner_any". key: 0, node: __cdfy/outfit_commoner_any"`;

exports[`prompts:error "prompts-bad/not-matched_003.yaml" 1`] = `"Invalid Syntax [UNSAFE_SYNTAX] "_" in value near "_cdfy/outfit_com". key: 0, node: _cdfy/outfit_commoner_any__"`;

exports[`prompts:error "prompts-bad/not-matched_004.yaml" 1`] = `"Invalid Syntax [UNSAFE_SYNTAX] "_" in value near "it_commoner_any_". key: 0, node: __cdfy/outfit_commoner_any_"`;
